tests/test_helpers/test_search/test_models.py

```diff
@@ -153,121 +160,217 @@ def test_constraint_type():
+    )
+    constraint_type_not_in_pb = constraint_type_not_in.encode()
+    actual_constraint_type_not_in = ConstraintType.decode(
+        constraint_type_not_in_pb, "set"
+    )
+    assert actual_constraint_type_not_in == constraint_type_not_in
 
-    constrant_type = ConstraintType(ConstraintTypes.DISTANCE, [Location(1.1, 2.2), 2.2])
-    constrant_type.is_valid(Attribute("test", int, True))
-    constrant_type.check(Location(1.1, 2.2))
+    constraint_location = (Location(1.1, 2.2), 2.2)
+    constraint_type_distance = ConstraintType(
+        ConstraintTypes.DISTANCE, constraint_location
+    )
+    constraint_type_distance.is_valid(Attribute("test", int, True))
+    constraint_type_distance.check(Location(1.1, 2.2))
+    assert constraint_type_distance == ConstraintType(
+        ConstraintTypes.DISTANCE, constraint_location
+    )
+    constraint_type_distance_pb = constraint_type_distance.encode()
+    actual_constraint_type_distance = ConstraintType.decode(
+        constraint_type_distance_pb, "distance"
+    )
+    assert actual_constraint_type_distance == constraint_type_distance
 
     with pytest.raises(ValueError):
         ConstraintType("something", [Location(1.1, 2.2), 2.2]).is_valid(
             Attribute("test", int, True)
         )
 
-    with pytest.raises(AEAEnforceError):
+    with pytest.raises(AEAEnforceError, match=""):
         ConstraintType(ConstraintTypes.GREATER_THAN, str)
 
-    assert ConstraintType(ConstraintTypes.IN, [1, 2]) == ConstraintType(
-        ConstraintTypes.IN, [1, 2]
-    )
+    list_value = [1, 2]
+    set_value = {1, 2}
+    list_location = [Location(1.1, 2.2), 2.2]
+
+    with pytest.raises(
+        AEAEnforceError, match=f"Expected tuple, got {type(list_value)}"
+    ):
+        ConstraintType(ConstraintTypes.WITHIN, list_value)
+
+    with pytest.raises(
+        AEAEnforceError, match=f"Expected tuple, got {type(list_value)}"
+    ):
+        ConstraintType(ConstraintTypes.IN, list_value)
+
+    with pytest.raises(AEAEnforceError, match=f"Expected tuple, got {type(set_value)}"):
+        ConstraintType(ConstraintTypes.IN, set_value)
+
+    with pytest.raises(
+        AEAEnforceError, match=f"Expected tuple, got {type(list_value)}"
+    ):
+        ConstraintType(ConstraintTypes.NOT_IN, list_value)
+
+    with pytest.raises(AEAEnforceError, match=f"Expected tuple, got {type(set_value)}"):
+        ConstraintType(ConstraintTypes.NOT_IN, set_value)
+
+    with pytest.raises(
+        AEAEnforceError, match=f"Expected tuple, got {type(list_location)}"
+    ):
+        ConstraintType(ConstraintTypes.DISTANCE, list_location)
 
 
 def test_constraints_expressions():
     """Test constraint expressions: And, Or, Not."""
-    expression = And(
+    and_expression = And(
         [
-            ConstraintType(ConstraintTypes.EQUAL, 12),
-            ConstraintType(ConstraintTypes.EQUAL, 12),
+            Constraint("number", ConstraintType(ConstraintTypes.GREATER_THAN, 15)),
```

Comment:
A side observation: The above is unintuitive. 

`Constraint("number", ConstraintType(ConstraintTypes.GREATER_THAN, 15))` intuitively (reading from left to right) means `number` must be greater than 15.

Currently it means 15 is greater than `number`.

All other constraint are read left to right (`within`, `in`, `not_in`, etc). But `>`, `>=`, `<`, and `<=` do not respect this. in the old `python-oef-sdk` repo, this is also respected.

To correct it, we need to modify lines 679-686 in `models.py` and just invert the operators.

Reply:
Good observation :). What's the effect of this proposed change on the rest of the code base?

Reply:
I thought the effect might be a lot at first, hence commented first before doing it.

After I did a search in the framework, apparently we do not use these operators much ourselves (as in skills, etc). 

Here is a rough list of where they are used:
- `soef` connection -> `object_translator.py`
- `tac` (might be one of its failure reasons)
- bunch of tests

But if there are other developers who do use these, meaning they were using concepts under `helpers.search` (which they could but weren't recommended anywhere to do so afaik), the semantics would gets reversed in those codes. However, mu hunch is if there were any, they would probably complain/report this oddity somewhere.

All in all, I suggest we do it in a separate PR to see all its effects. It would make this one messy if we do it here.

Reply:
Ok, let's do it but as you say in separate PR. We can make a note in upgrading guide re this breaking change.